File: src/295_Find_Median_from_Data_Stream/main.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class MedianFinder {
public:
    //Runtime: 128 ms, faster than 93.35%
    //Memory Usage: 46.7 MB, less than 97.41%
    /** initialize your data structure here. */
    MedianFinder() {
        // 对于每次传入一个未知的数，如果每次都排序，那么复杂度会变成n*O(nlogn)
        // 本题的难点在于需要尽快找到中间的两个值，其他的数只要分别在较小的组和较大的组就行了，详细的大小关系不用管
        // 注意，仅仅一个heap是搞不定的，因为help的规则只要父节点比子节点小就可以了，故无法保证节点是ordered的情况
    }

    void addNum(int num) {
        // 难点是加入一个数时要维护lhs和rhs
        if (lhs.empty()) {
            lhs.push(num);
            return; // 固定lhs的规模大于rhs
        }

        // 不管怎样都要先加入lhs洗礼一下
        lhs.push(num);
        // 把lhs最大的取出来，再放到rhs里面去洗礼一下
        rhs.push(lhs.top());
        // 好啦这样的话，左右分组继续满足左分组<=右分组的情况

        // 开始维护左右分组的数据规模，因为左右分支都增加了1，所以大小关系与先前并不变
        // e.g. [1,2] [50, 60]  + 100 -> [1,2,100] [50, 60, 100]，此时去掉rhs里面的top，将该top放入lhs
        // e.g. [1,15] [20]  + 10 -> [1,10,15] [15, 20]，此时去掉lhs里面的top
        // e.g. [1,15] [20]  + 100 -> [1,15,100] [20, 100]，此时去掉lhs里面的top
        if (lhs.size() == rhs.size()) {
            // 要把num放到lhs，保证比rhs多
            // lhs.top() = rhs.top();貌似还不允许这么操作
            lhs.push(rhs.top());
            lhs.pop();
            rhs.pop();
        } else {
            // 要把num放到rhs，去掉左边最大的那个值，那个值必定已经进入了rhs，
            lhs.pop();
        }
        return;
    }

    double findMedian() {
        double rtn = -1.0;
        if (lhs.empty()) return rtn;//因为默认lhs的数量比rhs多

        // 如果lhs和rhs的数据规模相同，那么就是取平均
        if (lhs.size() == rhs.size()) {
            rtn = lhs.top() + (rhs.top() - lhs.top())/2.0;
        } else {
            rtn = lhs.top();
        }
        return rtn;
    }

    priority_queue<int> lhs;//数据规模为n/2, 所有数据<= rhs
    priority_queue<int, vector<int>, greater<int>> rhs;//数据规模n/2,所有数据>= lhs里的数据
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

int main() {
  MedianFinder gua;
  vector<int> test;

  test = vector<int>{1,2,-1,3,-1};
  for (int i = 0; i < test.size() ; ++i) {
  //for (auto val : test) {
    if (test[i] >= 0)
      gua.addNum(test[i]);
    else
      cout << "expect value = " << ((i==2)?1.5:2.0) << ", actual value = " << gua.findMedian() << endl;
  }
  
  return 0;
}

File: src/295_Find_Median_from_Data_Stream/main.py

import bisect
import heapq

class MedianFinder:
    # 采用了priorityq，python默认是按照greater排序，即min heap; C++默认是less排序，是max heap
    def __init__(self):
        """
        invariant: len(left_heap) + 1 >= len(right_heap) >= len(left_heap)
        """
        self.left_heap = [] # max heap
        self.right_heap = [] # min heap

    def addNum(self, num: int) -> None:
        if len(self.left_heap) == len(self.right_heap):
            heapq.heappush(self.left_heap, - num)
            top = - heapq.heappop(self.left_heap)
            heapq.heappush(self.right_heap, top)
        else:
            heapq.heappush(self.right_heap, num)
            low = heapq.heappop(self.right_heap)
            heapq.heappush(self.left_heap, - low)

    def findMedian(self) -> float:
        if len(self.left_heap) == len(self.right_heap):
            first = - self.left_heap[0]
            second = self.right_heap[0]
            return (first + second) / 2.0
        else:
            return self.right_heap[0]


class MedianFinder2:
    #Runtime: 252 ms, faster than 27.00%
    #Memory Usage: 25.8 MB, less than 24.10%
    # 这里可以的改进是，不一定要让左右两个bisect保持完美的递增order
    # 只要保证左半的bisect个数>=右半的bisect个数，就可以了；然后就是看lhalf[-1]和rhalf[0]这两个中间值
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.lhalf = []
        self.rhalf = []

    def addNum(self, num: int) -> None:
        # 总是先加入到左边
        pos = bisect.insort_right(self.lhalf, num)
        
        # 然后保证左右两半的顺序
        if len(self.rhalf) and self.lhalf[-1] > self.rhalf[0]:
            bisect.insort_right(self.rhalf, self.lhalf[-1])
            self.lhalf.pop(-1)
        
        # 现在开始维护两个分支的数量大小
        if len(self.lhalf) > len(self.rhalf) + 1:
            bisect.insort_right(self.rhalf, self.lhalf[-1])
            self.lhalf.pop(-1)
        elif len(self.lhalf) < len(self.rhalf):
            bisect.insort_right(self.lhalf, self.rhalf[0])
            self.rhalf.pop(0)

    def findMedian(self) -> float:
        if not len(self.lhalf): return 0
        
        if len(self.lhalf) > len(self.rhalf):
            return self.lhalf[-1]
        else:
            return (self.lhalf[-1] + self.rhalf[0])/2

if __name__ == "__main__":
    gua = MedianFinder()
    nums = [2,7,9,4,4]

    for num in nums:
      gua.addNum(num)
      rtn = gua.findMedian()
      print(rtn, end = ", ")

    print("") # 不加这一行的话会在shell里面的最后出现一个%

